### PR TITLE
[Frontend][Tensorflow]add batch_dim support for gatherV2

### PR DESCRIPTION
**Pull request by zxy844288792**
Encounter a special cases when batch_dims=1 in gather() from centernet_hourglass_512x512_1 from tensorflow hub model zoo. 
Implement the batch_dims logic according to tensorflow implementation: https://www.tensorflow.org/api_docs/python/tf/gather
https://github.com/tensorflow/tensorflow/blob/5dcfc51118817f27fad5246812d83e5dccdc5f72/tensorflow/core/kernels/gather_op.cc

I have not added testcases for topi and relay since numpy does not have that attribute, I am open to see any suggestion for that. Test cases for tensorflow frontend parser have been added. 

Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

File: include/tvm/topi/transform.h [hunk 1]
```diff
@@ -854,46 +857,99 @@ inline Tensor sequence_mask(const Tensor& data, const Tensor& valid_length, doub
...
+  if (batch_dims_ != 0) {
+    ICHECK_GE(batch_dims_, -static_cast<int>(indices->shape.size())) << "batch_dims out of bounds";
+    ICHECK_LE(batch_dims_, indices->shape.size()) << "batch_dims out of bounds";
+
+    if (batch_dims_ < 0) {
+      batch_dims_ = indices->shape.size() + batch_dims_;
     }
+
+    ICHECK_LT(batch_dims_, a->shape.size()) << "batch_dims out of bounds";
+    ICHECK_GE(axis, batch_dims_) << "batch_dims must be less than or equal to axis";
```
Comment (comaniac) [inline include/tvm/topi/transform.h:881]:
```suggestion
    ICHECK_LE(batch_dims_, axis) << "batch_dims must be less than or equal to axis";
```

File: src/topi/transform.cc [hunk 2]
```diff
@@ -87,13 +87,15 @@ TVM_REGISTER_GLOBAL("topi.layout_transform").set_body([](TVMArgs args, TVMRetVal
...
-    int axis = args[2];
+  if (args.size() == 4) {
     std::string mode = args[3];
-    *rv = take(args[0], args[1], axis, mode);
+    int batch_dims = args[2];
+    *rv = take(args[0], args[1], batch_dims, mode);
+  } else {
+    int batch_dims = args[2];
+    int axis = args[3];
+    std::string mode = args[4];
```
Comment (comaniac) [inline src/topi/transform.cc:97]: So the nargs is expected to be 5? Please add a CHECK.

**Review by comaniac — APPROVED** [review]
Overall LGTM. Just nits.

File: src/topi/transform.cc (hunk 2 above)
Comment (zxy844288792) [reply src/topi/transform.cc:98]: Agree, addressed in the latest commit

**Review by yongwww — APPROVED** [review]
LGTM

**Comment by tkonolige** [comment]
@zxy844288792 I notice you haven't update the gradient for take to accept batch_dims. Could you either update the gradient or add a check that batch_dims = 0 in the gradient?

**Comment by zxy844288792** [comment]
> @zxy844288792 I notice you haven't update the gradient for take to accept batch_dims. Could you either update the gradient or add a check that batch_dims = 0 in the gradient?

I just added a check for batch_dims == 0 in the gradient

**Comment by comaniac** [comment]
Thanks @zxy844288792 @tkonolige @yongwww

**Pull request merged**